nupic/research/frameworks/dendrites/dendritic_weights.py

```diff
@@ -0,0 +1,85 @@
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+"""
+A simple implementation of dendrite weights. This combines the output from a (sparse)
+linear layer with the output from a set of dendritic segments.
+"""
+
+from nupic.research.frameworks.dendrites import DendriteSegments
+from nupic.torch.modules.sparse_weights import SparseWeights
+
+
+class DendriticWeights(SparseWeights):
+    """
+    This combines a DendriteSegments module with a SparseLinear module.
+    The output from the dendrite segments (shape of num_units x num_segments)
+    is applied to the output of of the linear weights (shape of num_units).
+    Thus, each linear output unit gets modulated by a set of dendritic segments.
+    """
+
+    def __init__(
+        self, module, num_segments, dim_context,
+        module_sparsity, dendrite_sparsity, dendrite_bias=None
+    ):
+        """
+        :param module: linear module from in-units to out-units
+        :param num_segments: number of dendrite segments per out-unit
+        :param dim_context: length of the context vector;
+                            the same context will be applied to each segment
+        :param module_sparsity: sparsity applied over linear module;
+        :param dendrite_sparsity: sparsity applied transformation per unit per segment
+        """
+        self.segments = None
+        super().__init__(module, sparsity=module_sparsity)
+
+        self.segments = DendriteSegments(
+            num_units=module.weight.shape[0],
+            num_segments=num_segments,
+            dim_context=dim_context,
+            sparsity=dendrite_sparsity,
+            bias=dendrite_bias,
+        )
+
+        self.rezero_weights()
+
+    def rezero_weights(self):
+        super().rezero_weights()
+        if self.segments is not None:  # only none at beggining of init
+            self.segments.rezero_weights()
+
+    def apply_dendrites(self, y, dendrite_activations):
+        """Apply dendrites as a bias."""
+        return y + dendrite_activations.max(dim=2).values  # max along each segment
+
+    def forward(self, x, context):
+        """
+        Compute of linear layer and apply output of dendrite segments.
+        """
+        y = super().forward(x)
+        dendrite_activations = self.segments(context)  # num_units x num_segments
+        return self.apply_dendrites(y, dendrite_activations)
+
+
+class GatingDendriticWeights(DendriticWeights):
```

Comment:
If I understand correctly,

`DendriticWeights` --> applies dendrite activations as bias,
`GatingDendriticWeights` --> applies dendrite activations as multiplicative value,

so why is one a subclass of the other?

Reply:
That's correct.

See this comment above:
https://github.com/numenta/nupic.research/pull/356#discussion_r497548307

I don't expect these classes to be the final way we organize the code, but a simple template to iterate from. The `apply_dendrites` function was something, in particular, I wanted to be easily changed.